### PR TITLE
Support platform messages in Linux shell

### PR DESCRIPTION
**Pull request by robert-ancell**

### CODE REVIEW

**Comment by robert-ancell** [comment]
This is a first iteration on the platform message support and allows sending of messages in both directions. It is built on top of #17910. This is the MVP of this feature, submitting here for feedback.

An example on these new APIs is:

```c
static FlView *view;

static void
message_cb (FlEngine *engine, const gchar *channel, GBytes *message, FlEnginePlatformMessageHandle *handle, gpointer user_data)
{
    g_autoptr(GBytes) response = do_something_with_message (message);

    g_autoptr(GError) error = NULL;
    if (!fl_engine_send_platform_message_response (engine, handle, response, &error))
        g_printerr ("Failed to send platform message response: %s\n", error->message);
}

static void
message_response_cb (GObject *object, GAsyncResult *result, gpointer user_data)
{
    g_autoptr(GError) error = NULL;
    g_autoptr(GBytes) response = fl_engine_send_platform_message_finish (fl_view_get_engine (view),
                                                                         result,
                                                                         &error);
   
    if (response == NULL) {
        g_printerr ("Failed to send platform message: %s\n", error->message);
        return;
    }

    do_something_with_response (response);
}

static void
some_external_event ()
{
    g_autoptr(GBytes) message = make_request ();

    fl_engine_send_platform_message (fl_view_get_engine (view),
                                     channel,
                                     message,
                                     NULL,
                                     message_response_cb,
                                     NULL);
}

int
main (int argc, char **argv)
{
    view = set_up_gtk ();

    g_autoptr(FlDartProject) project = fl_dart_project_new (".");
    view = fl_view_new (project);

    fl_engine_set_platform_message_handler (fl_view_get_engine (view),
                                            "example.com/gtk-test-method",
                                            message_cb,
                                            NULL);

    gtk_main ();

    return 0;
}
```

I've intentionally kept the APIs simple and as close to the GLib/GTK patterns. Do we then build the higher level codec/plugins on top of this or restructure it? Should we land this as working but refactor it later if the higher level code requires the API to change?

cc @stuartmorgan

File: shell/platform/linux/fl_engine.cc [hunk 1]
```diff
@@ -8,19 +8,108 @@
...
+
   FlDartProject* project;
   FlRenderer* renderer;
   FLUTTER_API_SYMBOL(FlutterEngine) engine;
+
+  // PlatformMessageHandler keyed by channel name
+  GHashTable* platform_message_handlers;
+
+  // Messages being processed
+  GPtrArray* message_handles;
```
Comment (stuartmorgan) [inline shell/platform/linux/fl_engine.cc:30]: What's the purpose of this array? I don't see any use of it other than adding and removing things. Is it just lifetime enforcement? If so, doesn't providing the handle to the handler take care of that?

File: shell/platform/linux/fl_engine.cc [hunk 2]
```diff
@@ -8,19 +8,108 @@
...
+  const FlutterPlatformMessage* message;
+  GBytes* data;
+};
+
+static FlEnginePlatformMessageHandle* message_handle_new(
+    const FlutterPlatformMessage* message) {
+  FlEnginePlatformMessageHandle* handle =
+      static_cast<FlEnginePlatformMessageHandle*>(
+          g_malloc0(sizeof(FlEnginePlatformMessageHandle)));
+  handle->message = message;
```
Comment (stuartmorgan) [inline shell/platform/linux/fl_engine.cc:74]: It looks like you're keeping a copy of a struct full of dangling pointers here. Anything you want to put into a reference-counted object from `message` needs to be copied.

File: shell/platform/linux/fl_renderer_x11.cc [hunk 3]
```diff
@@ -25,9 +25,12 @@ static void fl_renderer_x11_class_init(FlRendererX11Class* klass) {
 
 static void fl_renderer_x11_init(FlRendererX11* self) {}
 
-FlRendererX11* fl_renderer_x11_new(Window xid) {
-  FlRendererX11* self = static_cast<FlRendererX11*>(
+FlRendererX11* fl_renderer_x11_new() {
+  return static_cast<FlRendererX11*>(
```
Comment (stuartmorgan) [inline shell/platform/linux/fl_renderer_x11.cc:29]: It looks like some unrelated changes ended up in this PR?

File: shell/platform/linux/public/flutter_linux/fl_engine.h [hunk 4]
```diff
@@ -23,6 +24,99 @@ G_DECLARE_FINAL_TYPE(FlEngine, fl_engine, FL, ENGINE, GObject)
...
+/**
+ * fl_engine_set_platform_message_handler:
+ * @engine: a #FlEngine
+ * @channel: channel to listen on
+ * @callback: function to call when a message is received on this channel
+ * @user_data: (closure): user data to pass to @callback
+ *
+ * Register a callback to handle platform messages. Call
+ * fl_engine_send_platform_message_response() when this message should be
+ * responded to.
```
Comment (stuartmorgan) [inline shell/platform/linux/public/flutter_linux/fl_engine.h:60]: All messages should be responded to; failure to respond is a leak. The comment should make that very clear to help avoid people using it incorrectly.

File: shell/platform/linux/public/flutter_linux/fl_engine.h [hunk 5]
```diff
@@ -23,6 +24,99 @@ G_DECLARE_FINAL_TYPE(FlEngine, fl_engine, FL, ENGINE, GObject)
...
+ *
+ * Respond to a platform message.
+ *
+ * Returns: %TRUE on success.
+ */
+gboolean fl_engine_send_platform_message_response(
+    FlEngine* engine,
+    FlEnginePlatformMessageHandle* handle,
+    GBytes* response,
+    GError** error);
```
Comment (stuartmorgan) [inline shell/platform/linux/public/flutter_linux/fl_engine.h:84]: Why does sending a response have an `error` but sending a message doesn't? It seems like an odd asymmetry.

File: shell/platform/linux/public/flutter_linux/fl_engine.h [hunk 6]
```diff
@@ -23,6 +24,99 @@ G_DECLARE_FINAL_TYPE(FlEngine, fl_engine, FL, ENGINE, GObject)
...
+
+/**
+ * fl_engine_send_platform_message:
+ * @engine: a #FlEngine
+ * @channel: channel to send to
+ * @message: (allow-none): message buffer to send or %NULL for an empty message
+ * @cancellable: (allow-none): a #GCancellable or %NULL
```
Comment (stuartmorgan) [inline shell/platform/linux/public/flutter_linux/fl_engine.h:91]: What does this do? It's not clear to me how this interacts with the callback; I can see that they are used to make the task object, but what is it for? (I tried looking at the docs, but didn't see a high-level overview in my initial skim.)

File: shell/platform/linux/public/flutter_linux/fl_engine.h [hunk 7]
```diff
@@ -23,6 +24,99 @@ G_DECLARE_FINAL_TYPE(FlEngine, fl_engine, FL, ENGINE, GObject)
...
+                                     gpointer user_data);
+
+/**
+ * fl_engine_send_platform_message_finish:
+ * @engine: a #FlEngine
+ * @result: a #GAsyncResult
+ * @error: (allow-none): #GError location to store the error occurring, or %NULL
+ * to ignore.
+ *
+ * Complete request started with fl_engine_send_platform_message().
```
Comment (stuartmorgan) [inline shell/platform/linux/public/flutter_linux/fl_engine.h:112]: Is there a nutshell description of what the purpose of this pattern is? I can see how you're using it in the example code, but how else would it ever be called (such that it's useful to have as a method rather than just providing the thing this would return directly to the callback)?

File: shell/platform/linux/fl_engine.cc [hunk 8]
```diff
@@ -125,6 +282,106 @@ gboolean fl_engine_start(FlEngine* self, GError** error) {
...
+  if (result != kSuccess) {
+    g_set_error(error, fl_renderer_error_quark(), FL_RENDERER_ERROR_FAILED,
+                "Failed to send platorm message response");
+    return FALSE;
+  }
+
+  return TRUE;
+}
+
+G_MODULE_EXPORT void fl_engine_send_platform_message(
```
Comment (stuartmorgan) [inline shell/platform/linux/fl_engine.cc:326]: You're not going to want these APIs to be on the engine:
- Plugins need access to messaging, but shouldn't get access to the whole engine API.
- We recently had to fix retain cycle leaks in some of the retain-count-based embeddings due the fact that the object given to plugins was actually the engine under the hood.

You should make a messenger interface like the other platforms have, and then implement it with an object that is created by the engine and has a weak pointer to the engine. (If you want to do the second in a follow-up and file a bug about it that's fine, but you should do the first now since it changes the API and we may as well avoid thrashing the API when we already know what the desired API is.)

**Comment by stuartmorgan** [comment]
> cc @stuartmorgan

FYI now that you have project permissions you can just add me as a reviewer directly.

File: shell/platform/linux/fl_engine.cc (hunk 1 above)
Comment (robert-ancell) [reply shell/platform/linux/fl_engine.cc:30]: It was to keep ownership of the GBytes and the handle, but I agree, with further thought it's not really necessary. It does require the API user to take a reference to the GBytes, but that's probably not unreasonable.

I have been struggling with the way to pass the response handle to the FlEngine user - it seems unnecessary to wrap it inside a structure with one element. But the only way we can otherwise pass it is using a void pointer, which is ugly.

File: shell/platform/linux/fl_renderer_x11.cc (hunk 3 above)
Comment (robert-ancell) [reply shell/platform/linux/fl_renderer_x11.cc:29]: I had to change the renderer to set the X11 window ID after construction - we need the engine as soon as the FlView is created, so we can register any platform message callbacks. However we don't know the X11 window ID until later. So the FlRendererX11 needed updating for that. It doesn't feel great, as it would be nicer for everything to be set at construction. We could resolve this by making a new object for sending and receiving messages, but that makes the API user do more work for no apparent gain. I ended up going with this solution for now and consider changing it later if that new class ends up being useful.

File: shell/platform/linux/public/flutter_linux/fl_engine.h (hunk 5 above)
Comment (robert-ancell) [reply shell/platform/linux/public/flutter_linux/fl_engine.h:84]: `fl_engine_send_platform_message` is an asynchronous call, so the response is in the callback. This includes reporting of errors. This is the GLib convention. Note I have GCancellable objects there too to match the convention even though we don't make use of them. There are some APIs out there that didn't do this and they're pretty hard to retrofit...

File: shell/platform/linux/public/flutter_linux/fl_engine.h (hunk 6 above)
Comment (robert-ancell) [reply shell/platform/linux/public/flutter_linux/fl_engine.h:91]: GCancellable is the means that the caller can request the the asynchronous call is stopped before the callback has been received. GTask is just a convenience object to make async code easier to write. While we're not able to actually send the cancel through to the Dart code, if a user does cancel the callback will return the error G_IO_ERROR_CANCELLED, which is the expected behaviour and will stop them from processing the response. All GLib async calls should take a GCancellable.

File: shell/platform/linux/public/flutter_linux/fl_engine.h (hunk 7 above)
Comment (robert-ancell) [reply shell/platform/linux/public/flutter_linux/fl_engine.h:112]: This is the pattern used in GLib asynchronous code. The methods are normally called:

foo_async (object, args, cancellable, callback, user_data)
ReturnedData foo_sync (object, args, error)
ReturnedData foo_finish (object, result, error);

That should be recognisable to anyone familiar with GLib. In this case I dropped the _async suffix because we don't have a need for a _sync variant and it should be clear that the call is async. Other libraries do this.

Language bindings e.g. Vala hide away the three methods with their built in async/await operations. GLib doesn't use a Future type object for async.

File: shell/platform/linux/fl_engine.cc (hunk 2 above)
Comment (robert-ancell) [reply shell/platform/linux/fl_engine.cc:74]: It wasn't clear to me the ownership of `FlutterPlatformMessage` - I assumed that is is valid until `FlutterEngineSendPlatformMessageResponse` is called? Does the `uint8_t*` data also need to be copied?

**Comment by robert-ancell** [comment]
@stuartmorgan I've added FlBinaryMessenger and tried to match the MacOS behaviour as much as possible. The only remaining issue is if we need to copy the message buffer.

The example code using this new API is:

```c
static FlView *view;

static void
message_cb (FlBinaryMessenger *messenger, const gchar *channel, GBytes *message, FlBinaryMessengerResponseHandle *handle, gpointer user_data)
{
    g_autoptr(GBytes) response = do_something_with_message (message);

    g_autoptr(GError) error = NULL;
    if (!fl_binary_messenger_send_response (messenger, handle, response, &error))
        g_printerr ("Failed to send platform message response: %s\n", error->message);
}

static void
message_response_cb (GObject *object, GAsyncResult *result, gpointer user_data)
{
    g_autoptr(GError) error = NULL;
    g_autoptr(GBytes) response = fl_binary_messenger_send_on_channel_finish (FL_BINARY_MESSENGER(object),
                                                                             result,
                                                                             &error);
   
    if (response == NULL) {
        g_printerr ("Failed to send platform message: %s\n", error->message);
        return;
    }

    do_something_with_response (response);
}

static void
some_external_event ()
{
    g_autoptr(GBytes) message = make_request ();

    fl_binary_messenger_send_on_channel (fl_engine_get_binary_messenger(fl_view_get_engine (view)),
                                         channel,
                                         message,
                                         NULL,
                                         message_response_cb,
                                         NULL);
}

int
main (int argc, char **argv)
{
    view = set_up_gtk ();

    g_autoptr(FlDartProject) project = fl_dart_project_new (".");
    view = fl_view_new (project);

    fl_binary_messenger_set_message_handler_on_channel (fl_engine_get_binary_messenger(fl_view_get_engine (view)),
                                                        "example.com/gtk-test-method",
                                                        message_cb,
                                                        NULL);

    gtk_main ();

    return 0;
}
```

File: shell/platform/linux/fl_engine.cc (hunk 2 above)
Comment (stuartmorgan) [reply shell/platform/linux/fl_engine.cc]: Answered in Discord, but to capture here: the lifetime of parameters is generally only guaranteed a during a call. We can't assume that the message is valid until the response is provided.

(In the case of the message, it's very common for that data to only be needed synchronously during handling of the message receipt, even if the response is async, so it would be inefficient for the engine to always store it.)

**Comment by robert-ancell** [comment]
Updated to copy the message based on discussion on Discord.

File: shell/platform/linux/public/flutter_linux/fl_binary_messenger.h [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+/**
+ * FlBinaryMessengerCallback:
+ * @messenger: a #FlBinaryMessenger
+ * @channel: channel message received on
+ * @message: message content received from Dart
+ * @response_handle: a handle to respond to the message with
+ * @user_data: (closure): data provided when registering this callback
+ *
+ * Function called when platform messages are received.
```
Comment (stuartmorgan) [inline shell/platform/linux/public/flutter_linux/fl_binary_messenger.h:45]: This comment should also indicate that replying is mandatory; the more relevant places it's mentioned, the better.

File: shell/platform/linux/fl_binary_messenger.cc [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+G_MODULE_EXPORT gboolean fl_binary_messenger_send_response(
+    FlBinaryMessenger* self,
+    FlBinaryMessengerResponseHandle* response_handle,
+    GBytes* response,
+    GError** error) {
+  g_return_val_if_fail(FL_IS_BINARY_MESSENGER(self), FALSE);
+  g_return_val_if_fail(response_handle != nullptr, FALSE);
+
+  gboolean result = fl_engine_send_platform_message_response(
+      self->engine, response_handle->response_handle, response, error);
```
[the remark this answers is not in the document]
Comment (robert-ancell) [reply shell/platform/linux/fl_binary_messenger.cc:133]: Hah! I just removed the outstanding handles being stored in the previous code, I guess I'll add that back now :)

[the remark this answers is not in the document]
Comment (robert-ancell) [reply shell/platform/linux/fl_binary_messenger.cc:152]: https://github.com/flutter/flutter/issues/56262

**Comment by robert-ancell** [comment]
I've also been uneasy about the lack of tests - I've opened https://github.com/flutter/flutter/issues/56261 to track this.
